development/tools/pkg/jobguard/status.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// Init fetches pull request details and gathers essential pieces of information
+// TODO: Do we really need this? What about PULL_PULL_SHA env?
+func (f *StatusFetcher) Init() error {
+	prDetails, err := f.pullRequestDetails()
+	if err != nil {
+		return errors.Wrapf(err, "while getting pull request details for %d", f.cfg.PullRequestNumber)
+	}
+
+	commitSHA, err := f.pullRequestHeadSHA(prDetails)
+	if err != nil {
+		return errors.Wrapf(err, "while getting status URL for %d", f.cfg.PullRequestNumber)
+	}
+
+	f.commitSHA = commitSHA
+	return nil
+}
+
+// Do fetches statuses for a pull request
+func (f *StatusFetcher) Do() ([]Status, error) {
+	if f.commitSHA == "" {
+		return nil, errors.New("Commit SHA not fetched")
+	}
+
+	statuses, err := f.fetchStatuses(f.commitSHA)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while fetching status for %d", f.cfg.PullRequestNumber)
+	}
+
+	return statuses, nil
+}
+
+func (f *StatusFetcher) pullRequestDetails() (map[string]interface{}, error) {
+	prDetailsURL := fmt.Sprintf("%s/repos/%s/%s/pulls/%d", f.cfg.Origin, f.cfg.Owner, f.cfg.Repository, f.cfg.PullRequestNumber)
+
+	resp, err := f.client.Get(prDetailsURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while doing request to %s", prDetailsURL)
+	}
+
+	var result map[string]interface{}
+	err = json.NewDecoder(resp.Body).Decode(&result)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while decoding response from request to %s", prDetailsURL)
+	}
+	defer closeResponseBody(resp)
+
+	return result, nil
+}
+
+func (f *StatusFetcher) pullRequestHeadSHA(prDetails map[string]interface{}) (string, error) {
+	head, ok := prDetails["head"].(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("Incorrect type for head: %T", prDetails["head"])
+	}
+
+	sha, ok := head["sha"].(string)
+	if !ok {
+		return "", fmt.Errorf("Incorrect type for sha: %T", head["sha"])
+	}
+
+	return sha, nil
+}
+
+type statusResponse struct {
+	TotalCount int      `json:"total_count"`
+	Statuses   []Status `json:"statuses"`
+}
+
+func (f *StatusFetcher) fetchStatuses(commmitSHA string) ([]Status, error) {
+	url := fmt.Sprintf("%s/repos/%s/%s/commits/%s/status", f.cfg.Origin, f.cfg.Owner, f.cfg.Repository, commmitSHA)
+
+	var statuses []Status
+
+	pageNo := 1
+
+	for {
+		log.Printf("\tPaginating through statuses... Page number: %d\n", pageNo)
+		pageURL := fmt.Sprintf("%s?page=%d&per_page=100", url, pageNo)
+
+		resp, err := f.client.Get(pageURL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "while doing request to %s", url)
+		}
+
+		var result statusResponse
+		err = json.NewDecoder(resp.Body).Decode(&result)
+		if err != nil {
+			return nil, errors.Wrapf(err, "while decoding response from request to %s", url)
+		}
+
+		if len(result.Statuses) == 0 {
+			return nil, fmt.Errorf("Error while paginating on pages. No statuses found on page %d", pageNo)
+		}
+
+		statuses = append(statuses, result.Statuses...)
+
+		closeResponseBody(resp)
```

Comment:
The response body should be closed before every return statement in this function. Consider deferring the closeResponseBody call right after obtaining the response.